colossalai/utils/tensor_detector/tensor_detector.py

```diff
@@ -0,0 +1,132 @@
+                        for par_name, param in self.module.named_parameters():
+                            if param.requires_grad and param.grad.equal(obj.grad):
+                                name = par_name + ' (with grad)'
+                    # in the case of common tensor, we can't
+                    # but we can still marked it as tensor(with grad)
+                    # actually, we can, by recording the data and comparing
+                    # that requires much more memory
+                    if name == 'Tensor' and (obj.is_leaf or obj.retains_grad):
+                        if obj.grad is not None:
+                            name = name + ' (with grad)'
+
+                    self.tensors[id(obj)].append(name)
+                    self.tensors[id(obj)].append(obj.device)
+                    self.tensors[id(obj)].append(obj.shape)
+                    self.tensors[id(obj)].append(obj.requires_grad)
+                    self.tensors[id(obj)].append(obj.dtype)
+                    self.tensors[id(obj)].append(self.get_tensor_mem(obj))
+                # recorded the order we got the tensor
+                # by this we can guess the tensor easily
+                self.order.append(id(obj))
+                # recorded all different devices
+                if obj.device not in self.devices:
+                    self.devices.append(obj.device)
+
+    
+    def print_tensors_state(self):
+        template_format = '{:3s}{:<30s}{:>10s}{:>20s}{:>10s}{:>20s}{:>15s}'
+        print('-' * LINE_WIDTH)
+        print(template_format.format('  ', 'Tensor', 'device', 'shape', 'grad', 'dtype', 'Mem'))
+        print('-' * LINE_WIDTH)
+
+
+        for tensor_id in self.order:
+            print(template_format.format('+',
+                                        str(self.tensors[tensor_id][0]),
+                                        str(self.tensors[tensor_id][1]),
+                                        str(tuple(self.tensors[tensor_id][2])),
+                                        str(self.tensors[tensor_id][3]),
+                                        str(self.tensors[tensor_id][4]),
+                                        str(self.tensors[tensor_id][5])))
+        
+        # trace where is the detect()
+        locate_info = inspect.stack()[2]
+        locate_msg = '"' +  locate_info.filename + '" line ' + str(locate_info.lineno)
+        print('-' * LINE_WIDTH)
+        print(f"Detect Location: {locate_msg}")
+        for device in self.devices:
+            if device == torch.device('cpu'):
+                continue
+            gpu_mem_alloc = self.mem_format(torch.cuda.memory_allocated(device))
+            print(f"Totle GPU Memery Allocated on {device} is {gpu_mem_alloc}")
+        print('-' * LINE_WIDTH)
+        print('\n\n')
+    
+    def detect(self):
+        self.collect_tensors_state()
+        self.print_tensors_state()
+        self.tensors.clear()
+        self.pre_order = self.order.copy()
+        self.order = []
```

Comment:
did you clear the self.devices ?

Reply:
no, in fact, I don't think it is necessary. Seems the total devices would not change during running.

Reply:
OK